File: stripe_terminal/android/src/main/kotlin/mek/stripeterminal/mappings/ChargeMappings.kt
package mek.stripeterminal.mappings

import com.stripe.stripeterminal.external.models.Charge
import com.stripe.stripeterminal.external.models.PaymentMethodDetails
import mek.stripeterminal.api.ChargeApi
import mek.stripeterminal.api.ChargeStatusApi
import mek.stripeterminal.api.PaymentMethodDetailsApi
import mek.stripeterminal.api.toApi
import mek.stripeterminal.toHashMap

fun Charge.toApi(): ChargeApi {
    return ChargeApi(
        amount = amount,
        currency = currency!!,
        status =
            when (status) {
                "pending" -> ChargeStatusApi.PENDING
                "failed" -> ChargeStatusApi.FAILED
                "succeeded" -> ChargeStatusApi.SUCCEEDED
                else -> throw Error("Unsupported $status")
            },
        paymentMethodDetails = paymentMethodDetails?.toApi(),
        description = description!!,
        id = id,
        metadata = metadata?.toHashMap() ?: hashMapOf(),
        statementDescriptorSuffix = statementDescriptorSuffix,
        calculatedStatementDescriptor = calculatedStatementDescriptor,
        authorizationCode = authorizationCode,
    )
}

fun PaymentMethodDetails.toApi(): PaymentMethodDetailsApi {
    return PaymentMethodDetailsApi(
        cardPresent = cardPresentDetails?.toApi(),
        interacPresent = interacPresentDetails?.toApi(),
    )
}

File: stripe_terminal/ios/Classes/Mappings/ChargeMappings.swift
import Foundation
import StripeTerminal

extension ChargeStatus {
    func toApi() -> ChargeStatusApi {
        switch (self) {
        case ChargeStatus.pending:
            return ChargeStatusApi.pending
        case ChargeStatus.failed:
            return ChargeStatusApi.failed
        case ChargeStatus.succeeded:
            return ChargeStatusApi.succeeded
        @unknown default:
            fatalError("Unkown charge status")
        }
    }
}

extension Charge {
    func toApi() -> ChargeApi {
        return ChargeApi(
            amount: amount.intValue,
            authorizationCode: authorizationCode,
            calculatedStatementDescriptor: calculatedStatementDescriptor,
            currency: currency,
            description: description,
            id: stripeId,
            metadata: metadata,
            paymentMethodDetails: paymentMethodDetails?.toApi(),
            statementDescriptorSuffix: statementDescriptorSuffix,
            status: status.toApi()
        )
    }
}
